service/service.go

```diff
@@ -20,116 +20,218 @@ import (
+	mlpb "ml_metadata/proto/metadata_store_go_proto"
 
-	"github.com/golang/glog"
 	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes/empty"
 	pb "github.com/kubeflow/metadata/api"
-	mlmd "github.com/kubeflow/metadata/ml_metadata"
-	"github.com/kubeflow/metadata/schemaparser"
-)
-
-const (
-	wholeMetaPropertyName string = "__ALL_META__"
-	categoryPropertyName  string = "category"
-	artifactCategory      string = "artifact"
-	defaultTimeout               = 5 * time.Second
 )
 
 // Service implements the gRPC service MetadataService defined in the metadata
 // API spec.
 type Service struct {
-	mlmdClient       mlmd.MetadataStoreServiceClient
-	typeNameToMLMDID map[string]int64
+	store MetadataStore
+}
 
-	// TODO: Once MLMD type system supports schema annotations, the following two fields can be obtained by quering MLMD types.
-	// map from type name to the schema id, which is used to verify the incoming data.
-	typeNameToID map[string]string
-	schemaset    *schemaparser.SchemaSet
+// New returns a new instance of Service.
+func New(store MetadataStore) *Service {
+	return &Service{store: store}
 }
 
-// NewService returns a new MetadataService.
-func NewService(mlmdClient mlmd.MetadataStoreServiceClient, schemaRootDir string) (*Service, error) {
-	ss, err := schemaparser.NewSchemaSetFromADir(schemaRootDir)
-	if err != nil {
-		return nil, err
+// Close cleans up and frees resources held by Service.
+func (s *Service) Close() {
+	s.Close()
+}
+
+const (
+	kfWorkspace   = "kf_workspace"
+	kfNamespace   = "kf_namespace"
+	kfDescription = "kf_description"
+
+	kfDefaultNamespace = "types.kubeflow.org/default"
+)
+
+var validTypeNameRE = regexp.MustCompile(`^[A-Za-z][^ /]*$`)
+
+func validTypeName(n string) error {
+	if validTypeNameRE.MatchString(n) {
+		return nil
 	}
-	typeNameToID := make(map[string]string)
-	typeNameToMLMDID := make(map[string]int64)
-	for id := range ss.Schemas {
-		tn, err := ss.TypeName(id)
-		if err != nil {
-			glog.Warningf("schema misses 'kind', 'apiversion', or 'namespace'. schema $id = %s", id)
-			continue
-		}
-		category, err := ss.ConstantStringType(id, categoryPropertyName)
-		if err != nil {
-			return nil, fmt.Errorf("schema %q property. schema $id = %s", categoryPropertyName, id)
+
+	return fmt.Errorf("invalid type name %q: type names must begin with an alphabet and not contain spaces or slashes", n)
+}
+
+var validNamespaceRE = regexp.MustCompile(`^[A-Za-z][^ ]*[^/]$`)
+
+func validNamespace(n string) error {
+	if validNamespaceRE.MatchString(n) {
+		return nil
+	}
+
+	return fmt.Errorf("invalid namespace %q: namespaces must begin with an alphabet, should not contain spaces, or end with a trailing /", n)
+}
+
+func toMLMDArtifactType(in *pb.ArtifactType) (*mlpb.ArtifactType, error) {
+	res := &mlpb.ArtifactType{
+		Id:         proto.Int64(in.Id),
+		Properties: make(map[string]mlpb.PropertyType),
+	}
+
+	for k, v := range in.TypeProperties {
+		if strings.HasPrefix(k, "kf_") {
```

Comment:
can we use `__kf_` as the prefix to avoid possible collisions?

Reply:
Done.